### PR TITLE
Add a new k8s environ provider

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

k8s gains an environ provider implementation. This is necessary to allow various credential related CLI commands to function, as well as paving the way for model deletion and pluggable CAAS providers like we do for clouds.

A consequence is that the caas agent facade no longer needs a bespoke Model() API and is reduced to being built from existing facade api plugins, allowing a bunch of code to be deleted.

## QA steps

bootstrap a cloud
deploy k8s
deploy a workload on top of k8s
ensure the various credential CLI works
ensure kill-controller etc work without errors

### CODE REVIEW

**Comment by nskaggs** [comment]
Looking good.

```
$ juju list-credentials --format yaml
local-credentials:
  k8cloud:
    admin:
      auth-type: userpass
      Username: admin

$ juju remove-credential k8cloud admin
Local credential "admin" for cloud "k8cloud" has been deleted.

$ juju autoload-credentials

Looking for cloud and credential information locally...

1. kubernetes credential "admin" (new)
2. LXD credential "localhost" (new)
Select a credential to save by number, or type Q to quit: 1

Select the cloud it belongs to, or type Q to quit [k8cloud]: 

Saved kubernetes credential "admin" to cloud k8cloud

1. kubernetes credential "admin" (existing, will overwrite)
2. LXD credential "localhost" (new)
Select a credential to save by number, or type Q to quit: q
```

File: apiserver/facades/agent/caasagent/caasagent.go [hunk 1]
```diff
@@ -9,21 +9,22 @@ import (
...
 	cloudspec.CloudSpecAPI
-
-	model Model
+	*common.ModelWatcher
 }
 
 // NewStateFacade provides the signature required for facade registration.
 func NewStateFacade(ctx facade.Context) (*Facade, error) {
 	authorizer := ctx.Auth()
+	if !authorizer.AuthMachineAgent() {
```
Comment (nskaggs) [inline apiserver/facades/agent/caasagent/caasagent.go]: When would this trigger, or is this just a catch all?

File: caas/kubernetes/provider/credentials.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import (
+	"github.com/juju/errors"
+	"github.com/juju/juju/caas/kubernetes/clientconfig"
+
+	"github.com/juju/juju/cloud"
+	"github.com/juju/juju/environs"
+)
+
+const (
+	CredAttrUserName = "Username"
```
Comment (nskaggs) [inline caas/kubernetes/provider/credentials.go]: Perhaps match upstream, "username" and "password". Also CredAttrUserName or CredAttrUsername ?

File: cmd/juju/controller/kill.go [hunk 3]
```diff
@@ -199,21 +199,30 @@ func (c *killCommand) DirectDestroyRemaining(ctx *cmd.Context, api destroyContro
...
 			logger.Errorf(err.Error())
 			hasErrors = true
 			continue
 		}
-		if err := env.Destroy(); err != nil {
-			logger.Errorf(err.Error())
-			hasErrors = true
-		} else {
-			ctx.Infof("  done")
+		// TODO(caas) - only cloud providers support Destroy()
```
Comment (nskaggs) [inline cmd/juju/controller/kill.go]: What would CAAS support?

Comment (wallyworld) [reply cmd/juju/controller/kill.go]: CAAS would also ultimately have something similar

File: caas/kubernetes/provider/credentials.go (hunk 2 above)
Comment (wallyworld) [reply caas/kubernetes/provider/credentials.go]: CredAttrUserName was copied from elsewhere so I didn't change it

File: apiserver/facades/agent/caasagent/caasagent.go (hunk 1 above)
Comment (wallyworld) [reply apiserver/facades/agent/caasagent/caasagent.go]: It's a catch all - our standard way of excluding other agent types

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**